### ISSUE TITLE
Question: Expanding the ERA Split Logic

### ISSUE THREAD

**Issue by simpsus** [jefferythewind/scikit-learn-erasplit#4]
Hi JefferyTheWind,

I want to experiment by expanding the era splitting criterion and I wonder where the right place to implement that is.
The current implementation defines the era wise gain as the mean over all eras

![image](https://github.com/jefferythewind/scikit-learn-erasplit/assets/1447101/51092776-14c9-4178-b7d6-f115ebcb3027)

And then the Boltzmann operator afterwards. The boltzmann can stay as it is, but I want to experiment on the mean of all eras.
Can you point me to where this is implemented?
Is it this file

https://github.com/jefferythewind/scikit-learn-erasplit/blob/era_splitting/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx

Or even the histogram?

https://github.com/jefferythewind/scikit-learn-erasplit/blob/era_splitting/sklearn/ensemble/_hist_gradient_boosting/era_histogram.pyx

Thx!

**Comment by jefferythewind** [comment]
Hi, awesome thank you. Yes you were right in the first script, although the histogram logic was also changed but  probably you don't need to look there first. 

On this line is were we compile the pieces into the new split gain value. 

https://github.com/jefferythewind/scikit-learn-erasplit/blob/79f505b014724c88bed7d218816539e782fa7855/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx#L980

The key to era splitting is that we need to computer a gain value for each era, which loop starts here:

https://github.com/jefferythewind/scikit-learn-erasplit/blob/79f505b014724c88bed7d218816539e782fa7855/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx#L842

For each era we compute an era_gain for each era and make an average using the Boltzmann operator. But looking at it now it seems like there maybe a problem on this line. Did I mess up an indentation here?

https://github.com/jefferythewind/scikit-learn-erasplit/blob/79f505b014724c88bed7d218816539e782fa7855/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx#L918


This is all part of a "Left to Right" pass through the data, originally there was also a function for "Right to Left" pass, but I took it out to save time and it didn't seem to affect results in my tests. 

Best,

**Comment by simpsus** [comment]
The way I see the identation in 

https://github.com/jefferythewind/scikit-learn-erasplit/blob/79f505b014724c88bed7d218816539e782fa7855/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx#L918

the addition to the nominator&denominator happens outside of the loop starting in 

https://github.com/jefferythewind/scikit-learn-erasplit/blob/79f505b014724c88bed7d218816539e782fa7855/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx#L842

Now this is the first pyrex file I am looking at and I have read but not inhaled your paper ... my hunch is this "buildup" of the nominator & denominator should happen inside the loop, as it should aggregate the values for each era. The way it is now ... I wonder why era_gain is even defined and does not throw an error.

For my endavour, era_gain probably is the answer as well. I want to pass a weight-mask to the method which emphasizes a few eras over others. So I want to give each era_gain a factor of 1 by default and some another one.

Is era_gain the place to do that?
How would I pass an array of weights (one factor per era) to this method?

Thank you very much!

**Comment by simpsus** [comment]
I tried to test the loop-correction (the identation of the era gain addition), but the era_splitting file is a .pyd so I cannot edit that without recompiling the whole sklearn package

**Comment by jefferythewind** [comment]
Hi Sorry, I had commented previously but I think that the comment didn't go through. You should see in the most recent commit that I fixed the indentation error so if you re-pull the latest and re-install it should fix it. The fix is in a .pyx file, so yes you have to re-compile for the fixes to take affect. Re-installing with pip will do this re-compilation for you.

https://github.com/jefferythewind/scikit-learn-erasplit/blob/cf90868f86f606c87d20f4ad7dc598b049f47a2c/sklearn/ensemble/_hist_gradient_boosting/era_splitting.pyx#L918

**Comment by simpsus** [comment]
awesome, thank you very much!
I will close the issues and keep you posted on my development.
